File: android/src/main/java/app/capgo/rudderstack/RudderStackBridge.java
package app.capgo.rudderstack;

import android.content.Context;
import androidx.annotation.Nullable;
import com.getcapacitor.JSObject;
import com.rudderstack.android.sdk.core.RudderClient;
import com.rudderstack.android.sdk.core.RudderConfig;
import com.rudderstack.android.sdk.core.RudderOption;
import com.rudderstack.android.sdk.core.RudderProperty;
import com.rudderstack.android.sdk.core.RudderTraits;
import com.rudderstack.android.sdk.core.RudderTraitsBuilder;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import org.json.JSONArray;
import org.json.JSONObject;

final class RudderStackBridge {

    private final Context context;
    private @Nullable RudderClient client;

    RudderStackBridge(final Context context) {
        this.context = context.getApplicationContext();
    }

    void initialize(final String writeKey, @Nullable final JSObject configJson, @Nullable final JSObject optionsJson) {
        final RudderConfig config = getRudderConfig(configJson);
        final RudderOption options = getRudderOption(optionsJson);
        client = RudderClient.getInstance(context, writeKey, config, options);

        if (client == null) {
            throw new IllegalStateException("Failed to initialize RudderStack SDK");
        }
    }

    void identify(final String userId, @Nullable final JSObject traitsJson, @Nullable final JSObject optionsJson) {
        requireClient().identify(userId, getRudderTraits(traitsJson), getRudderOption(optionsJson));
    }

    void group(final String groupId, @Nullable final JSObject traitsJson, @Nullable final JSObject optionsJson) {
        requireClient().group(groupId, getRudderTraits(traitsJson), getRudderOption(optionsJson));
    }

    void track(final String eventName, @Nullable final JSObject propertiesJson, @Nullable final JSObject optionsJson) {
        requireClient().track(eventName, getRudderProperty(propertiesJson), getRudderOption(optionsJson));
    }

    void screen(final String screenName, @Nullable final JSObject propertiesJson, @Nullable final JSObject optionsJson) {
        requireClient().screen(screenName, getRudderProperty(propertiesJson), getRudderOption(optionsJson));
    }

    void alias(final String newId, @Nullable final JSObject optionsJson) {
        requireClient().alias(newId, getRudderOption(optionsJson));
    }

    void reset() {
        requireClient().reset();
    }

    void flush() {
        requireClient().flush();
    }

    void putDeviceToken(final String deviceToken) {
        RudderClient.putDeviceToken(deviceToken);
    }

    void putAdvertisingId(final String advertisingId) {
        RudderClient.putAdvertisingId(advertisingId);
    }

    void putAnonymousId(final String anonymousId) {
        RudderClient.putAnonymousId(anonymousId);
    }

    void optOut(final boolean optOut) {
        requireClient().optOut(optOut);
    }

    private RudderClient requireClient() {
        if (client == null) {
            throw new IllegalStateException("RudderStack SDK is not initialized");
        }

        return client;
    }

    private RudderConfig getRudderConfig(@Nullable final JSONObject configJson) {
        if (configJson == null) {
            return new RudderConfig.Builder().build();
        }

        final Map<String, Object> configMap = getMapFromJSON(configJson);
        final RudderConfig.Builder configBuilder = new RudderConfig.Builder();

        if (configMap.containsKey("dataPlaneUrl") && getString(configMap.get("dataPlaneUrl")) != null) {
            configBuilder.withDataPlaneUrl(getString(configMap.get("dataPlaneUrl")));
        }
        if (configMap.containsKey("flushQueueSize") && getInteger(configMap.get("flushQueueSize")) != null) {
            configBuilder.withFlushQueueSize(getInteger(configMap.get("flushQueueSize")));
        }
        if (configMap.containsKey("dbCountThreshold") && getInteger(configMap.get("dbCountThreshold")) != null) {
            configBuilder.withDbThresholdCount(getInteger(configMap.get("dbCountThreshold")));
        }
        if (configMap.containsKey("configRefreshInterval") && getInteger(configMap.get("configRefreshInterval")) != null) {
            configBuilder.withConfigRefreshInterval(getInteger(configMap.get("configRefreshInterval")));
        }
        if (configMap.containsKey("logLevel") && getInteger(configMap.get("logLevel")) != null) {
            configBuilder.withLogLevel(getInteger(configMap.get("logLevel")));
        }
        if (configMap.containsKey("sleepTimeOut") && getInteger(configMap.get("sleepTimeOut")) != null) {
            configBuilder.withSleepCount(getInteger(configMap.get("sleepTimeOut")));
        }
        if (configMap.containsKey("autoCollectAdvertId")) {
            configBuilder.withAutoCollectAdvertId(getBoolean(configMap.get("autoCollectAdvertId")));
        }
        if (configMap.containsKey("trackLifecycleEvents")) {
            configBuilder.withTrackLifecycleEvents(getBoolean(configMap.get("trackLifecycleEvents")));
        }
        if (configMap.containsKey("recordScreenViews")) {
            configBuilder.withRecordScreenViews(getBoolean(configMap.get("recordScreenViews")));
        }
        if (configMap.containsKey("controlPlaneUrl") && getString(configMap.get("controlPlaneUrl")) != null) {
            configBuilder.withControlPlaneUrl(getString(configMap.get("controlPlaneUrl")));
        }

        return configBuilder.build();
    }

    private @Nullable RudderOption getRudderOption(@Nullable final JSONObject optionsJson) {
        if (optionsJson == null) {
            return null;
        }

        final Map<String, Object> optionsMap = getMapFromJSON(optionsJson);
        final RudderOption option = new RudderOption();

        if (optionsMap.containsKey("externalIds")) {
            final Map<String, Object> externalIdsMap = castMap(optionsMap.get("externalIds"));
            for (Map.Entry<String, Object> entry : externalIdsMap.entrySet()) {
                final String externalId = getString(entry.getValue());
                if (externalId != null) {
                    option.putExternalId(entry.getKey(), externalId);
                }
            }
        }

        if (optionsMap.containsKey("integrations")) {
            final Map<String, Object> integrationsMap = castMap(optionsMap.get("integrations"));
            for (Map.Entry<String, Object> entry : integrationsMap.entrySet()) {
                option.putIntegration(entry.getKey(), getBoolean(entry.getValue()));
            }
        }

        return option;
    }

    private @Nullable RudderTraits getRudderTraits(@Nullable final JSONObject traitsJson) {
        if (traitsJson == null) {
            return null;
        }

        final Map<String, Object> traitsMap = getMapFromJSON(traitsJson);
        final RudderTraits traits = new RudderTraitsBuilder().build();
        for (Map.Entry<String, Object> entry : traitsMap.entrySet()) {
            traits.put(entry.getKey(), entry.getValue());
        }
        return traits;
    }

    private @Nullable RudderProperty getRudderProperty(@Nullable final JSONObject propertiesJson) {
        if (propertiesJson == null) {
            return null;
        }

        return new RudderProperty().putValue(getMapFromJSON(propertiesJson));
    }

    private Map<String, Object> getMapFromJSON(final JSONObject jsonObject) {
        final Map<String, Object> map = new HashMap<>();
        final Iterator<String> keysIter = jsonObject.keys();
        while (keysIter.hasNext()) {
            final String key = keysIter.next();
            final Object value = jsonObject.isNull(key) ? null : getObject(jsonObject.opt(key));
            if (value != null) {
                map.put(key, value);
            }
        }
        return map;
    }

    private Object getObject(final Object value) {
        if (value instanceof JSONObject) {
            return getMapFromJSON((JSONObject) value);
        }
        if (value instanceof JSONArray) {
            return getListFromJSON((JSONArray) value);
        }
        return value;
    }

    private @Nullable String getString(final Object value) {
        if (value instanceof String) {
            return (String) value;
        }
        if (value instanceof Number) {
            return String.valueOf(value);
        }
        return null;
    }

    private @Nullable Integer getInteger(final Object value) {
        if (value instanceof Number) {
            return ((Number) value).intValue();
        }
        if (value instanceof String) {
            return Integer.valueOf((String) value);
        }
        return null;
    }

    private boolean getBoolean(final Object value) {
        if (value instanceof Boolean) {
            return (Boolean) value;
        }
        if (value instanceof Number) {
            return ((Number) value).intValue() > 0;
        }
        if (value instanceof String) {
            return Boolean.parseBoolean((String) value);
        }
        return false;
    }

    private List<Object> getListFromJSON(final JSONArray jsonArray) {
        final List<Object> list = new ArrayList<>();
        for (int i = 0, count = jsonArray.length(); i < count; i++) {
            final Object value = getObject(jsonArray.opt(i));
            if (value != null) {
                list.add(value);
            }
        }
        return list;
    }

    @SuppressWarnings("unchecked")
    private Map<String, Object> castMap(final Object value) {
        if (value instanceof Map) {
            return (Map<String, Object>) value;
        }
        return new HashMap<>();
    }
}

File: ios/Sources/RudderStackPlugin/RudderStackBridge.swift
import Foundation
import Rudder

enum RudderStackBridgeError: LocalizedError {
    case initializationFailed
    case notInitialized

    var errorDescription: String? {
        switch self {
        case .initializationFailed:
            return "Failed to initialize RudderStack SDK"
        case .notInitialized:
            return "RudderStack SDK is not initialized"
        }
    }
}

final class RudderStackBridge {
    private let pluginVersion = "8.0.0"
    private var initialized = false

    var version: String {
        pluginVersion
    }

    func initialize(writeKey: String, config: [String: Any]?, options: [String: Any]?) throws {
        let rudderConfig = buildConfig(writeKey: writeKey, from: config)
        let rudderOptions = buildOptions(from: options)
        let client = RSClient.sharedInstance()
        client.configure(with: rudderConfig)
        if let rudderOptions {
            client.setOption(rudderOptions)
        }
        initialized = true
    }

    func identify(userId: String, traits: [String: Any]?, options: [String: Any]?) throws {
        let client = try requireClient()
        let option = buildOptions(from: options)

        if let option, let traits {
            client.identify(userId, traits: traits, option: option)
        } else if let traits {
            client.identify(userId, traits: traits)
        } else if let option {
            client.identify(userId, traits: [:], option: option)
        } else {
            client.identify(userId)
        }
    }

    func group(groupId: String, groupTraits: [String: Any]?, options: [String: Any]?) throws {
        let client = try requireClient()
        let option = buildOptions(from: options)
        let traits = stringifyMap(groupTraits)

        if let option, let traits {
            client.group(groupId, traits: traits, option: option)
        } else if let traits {
            client.group(groupId, traits: traits)
        } else if let option {
            client.group(groupId, traits: [:], option: option)
        } else {
            client.group(groupId)
        }
    }

    func track(eventName: String, properties: [String: Any]?, options: [String: Any]?) throws {
        let client = try requireClient()
        let option = buildOptions(from: options)

        if let option, let properties {
            client.track(eventName, properties: properties, option: option)
        } else if let properties {
            client.track(eventName, properties: properties)
        } else if let option {
            client.track(eventName, properties: [:], option: option)
        } else {
            client.track(eventName)
        }
    }

    func screen(screenName: String, properties: [String: Any]?, options: [String: Any]?) throws {
        let client = try requireClient()
        let option = buildOptions(from: options)

        if let option, let properties {
            client.screen(screenName, category: "", properties: properties, option: option)
        } else if let properties {
            client.screen(screenName, properties: properties)
        } else if let option {
            client.screen(screenName, category: "", properties: [:], option: option)
        } else {
            client.screen(screenName)
        }
    }

    func alias(newId: String, options: [String: Any]?) throws {
        let client = try requireClient()
        if let option = buildOptions(from: options) {
            client.alias(newId, option: option)
        } else {
            client.alias(newId)
        }
    }

    func reset() throws {
        try requireClient().reset()
    }

    func flush() throws {
        try requireClient().flush()
    }

    func putDeviceToken(_ deviceToken: String) {
        RSClient.sharedInstance().setDeviceToken(deviceToken)
    }

    func putAdvertisingId(_ advertisingId: String) {
        RSClient.sharedInstance().setAdvertisingId(advertisingId)
    }

    func putAnonymousId(_ anonymousId: String) {
        RSClient.sharedInstance().setAnonymousId(anonymousId)
    }

    func optOut(_ optOut: Bool) throws {
        try requireClient().setOptOutStatus(optOut)
    }

    private func requireClient() throws -> RSClient {
        guard initialized else {
            throw RudderStackBridgeError.notInitialized
        }
        return RSClient.sharedInstance()
    }

    private func buildConfig(writeKey: String, from config: [String: Any]?) -> RSConfig {
        let configuration = RSConfig(writeKey: writeKey)

        if let value = string(from: config?["dataPlaneUrl"]) {
            configuration.dataPlaneURL(value)
        }
        if let value = int(from: config?["flushQueueSize"]) {
            configuration.flushQueueSize(value)
        }
        if let value = int(from: config?["dbCountThreshold"]) {
            configuration.dbCountThreshold(value)
        }
        if let value = int(from: config?["sleepTimeOut"]) {
            configuration.sleepTimeOut(value)
        }
        if let value = int(from: config?["logLevel"]), let logLevel = RSLogLevel(rawValue: value) {
            configuration.loglevel(logLevel)
        }
        if let value = bool(from: config?["trackLifecycleEvents"]) {
            configuration.trackLifecycleEvents(value)
        }
        if let value = bool(from: config?["recordScreenViews"]) {
            configuration.recordScreenViews(value)
        }
        if let value = string(from: config?["controlPlaneUrl"]) {
            configuration.controlPlaneURL(value)
        }

        return configuration
    }

    private func buildOptions(from options: [String: Any]?) -> RSOption? {
        guard let options else {
            return nil
        }

        let rudderOption = RSOption()

        if let externalIds = options["externalIds"] as? [String: Any] {
            for (key, value) in externalIds {
                if let id = string(from: value) {
                    rudderOption.putExternalId(key, withId: id)
                }
            }
        }

        if let integrations = options["integrations"] as? [String: Any] {
            for (key, value) in integrations {
                if let enabled = bool(from: value) {
                    rudderOption.putIntegration(key, isEnabled: enabled)
                }
            }
        }

        return rudderOption
    }

    private func string(from value: Any?) -> String? {
        switch value {
        case let value as String:
            return value
        case let value as NSNumber:
            return value.stringValue
        default:
            return nil
        }
    }

    private func int(from value: Any?) -> Int? {
        switch value {
        case let value as NSNumber:
            return value.intValue
        case let value as String:
            return Int(value)
        default:
            return nil
        }
    }

    private func bool(from value: Any?) -> Bool? {
        switch value {
        case let value as Bool:
            return value
        case let value as NSNumber:
            return value.boolValue
        case let value as String:
            return Bool(value)
        default:
            return nil
        }
    }

    private func stringifyMap(_ value: [String: Any]?) -> [String: String]? {
        guard let value else {
            return nil
        }

        var result: [String: String] = [:]
        for (key, rawValue) in value {
            if let stringValue = string(from: rawValue) {
                result[key] = stringValue
            }
        }

        return result
    }
}
